service/history/api/multioperation/api.go

```diff
@@ -0,0 +1,167 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package multioperation
+
+import (
+	"context"
+
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/api/matchingservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/service/history/api"
+	"go.temporal.io/server/service/history/api/startworkflow"
+	"go.temporal.io/server/service/history/api/updateworkflow"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+func Invoke(
+	ctx context.Context,
+	req *historyservice.ExecuteMultiOperationRequest,
+	shardContext shard.Context,
+	workflowConsistencyChecker api.WorkflowConsistencyChecker,
+	tokenSerializer common.TaskTokenSerializer,
+	visibilityManager manager.VisibilityManager,
+	matchingClient matchingservice.MatchingServiceClient,
+) (_ *historyservice.ExecuteMultiOperationResponse, retError error) {
+	namespaceEntry, err := api.GetActiveNamespace(shardContext, namespace.ID(req.GetNamespaceId()))
+	if err != nil {
+		return nil, err
+	}
+	namespaceID := namespaceEntry.ID()
+
+	startReq := req.Operations[0].GetStartWorkflow()
+	updateReq := req.Operations[1].GetUpdateWorkflow()
+
+	// grab current Workflow lease, if there is one
+	currentWorkflowLease, err := workflowConsistencyChecker.GetWorkflowLease(
+		ctx,
+		nil,
+		api.BypassMutableStateConsistencyPredicate,
+		definition.NewWorkflowKey(
+			string(namespaceID),
+			startReq.StartRequest.WorkflowId,
+			"",
+		),
+		workflow.LockPriorityHigh,
+	)
+	switch err.(type) {
+	case nil:
+		defer func() { currentWorkflowLease.GetReleaseFn()(retError) }()
+	case *serviceerror.NotFound:
+		currentWorkflowLease = nil
+	default:
+		return nil, err
+	}
+
+	if currentWorkflowLease != nil {
+		return nil, serviceerror.NewUnimplemented("not implemented yet")
+	}
+
+	starter, err := startworkflow.NewStarter(
+		shardContext,
+		workflowConsistencyChecker,
+		tokenSerializer,
+		visibilityManager,
+		startReq,
+	)
+	if err != nil {
+		// TODO: send per-operation error details
+		return nil, err
+	}
+
+	var workflowCtx workflow.Context
+	var updateErr error
+	updater := updateworkflow.NewUpdater(
+		shardContext,
+		workflowConsistencyChecker,
+		matchingClient,
+		updateReq,
+	)
+	startResp, err := starter.Invoke(
+		ctx,
+		func(lease api.WorkflowLease) error {
+			workflowCtx = lease.GetContext()
+			_, updateErr = updater.ApplyRequest(ctx, lease)
+			return updateErr
+		},
+	)
+	if err != nil {
+		// TODO: send per-operation error details
+		return nil, err
```

Comment:
Could you provide at least a little context here while you refine this? The same goes for the start operation above

```suggestion
		return nil, fmt.Errorf("failed to update workflow: %w", err)
```